File: apps/ios/Moodish/Moodish/App/AppState.swift
import Foundation

enum BootstrapPhase: Equatable {
    case connecting(attempt: Int)
    case ready
    case failed(String)
}

/// Root app state: who's logged in, server config, and the slow-reconnect
/// dance the web app does against a Render cold start (see app.js `boot()`).
@Observable
@MainActor
final class AppState {
    let sessionStore = SessionStore()
    let themeStore = ThemeStore()
    let api: APIClient

    var phase: BootstrapPhase = .connecting(attempt: 1)
    var authConfig: AuthConfig?
    var user: MoodishUser?
    var mealMemory: [MealMemoryEntry] = []
    var health: HealthStatus?

    var isLoggedIn: Bool { user != nil }

    init() {
        api = APIClient(sessionStore: sessionStore)
    }

    func bootstrap() async {
        let maxAttempts = 6
        var attempt = 1
        while attempt <= maxAttempts {
            phase = .connecting(attempt: attempt)
            do {
                let result = try await api.bootstrap()
                authConfig = result.config
                user = result.user
                mealMemory = result.mealMemory
                health = result.health
                phase = .ready
                return
            } catch {
                if attempt == maxAttempts {
                    phase = .failed("Moodish is taking longer than usual to wake up. Pull to retry.")
                    return
                }
                let delayMs = min(5000, 300 * (1 << attempt))
                try? await Task.sleep(nanoseconds: UInt64(delayMs) * 1_000_000)
                attempt += 1
            }
        }
    }

    func loginWithDemo() async throws {
        let response = try await api.demoLogin()
        user = response.user
    }

    func loginWithGoogleToken(_ token: String) {
        sessionStore.setSessionToken(token)
        Task { await bootstrap() }
    }

    func logout() async {
        try? await api.logout()
        sessionStore.logout()
        user = nil
        mealMemory = []
    }
}

File: apps/moodish_android/lib/app_state.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'core/api_client.dart';
import 'core/session_store.dart';
import 'core/models/auth_models.dart';

enum BootstrapPhase { connecting, ready, failed }

class AppState extends ChangeNotifier {
  final ApiClient api = ApiClient();
  final SessionStore sessionStore = SessionStore();

  BootstrapPhase phase = BootstrapPhase.connecting;
  int connectingAttempt = 1;
  String? failureMessage;

  AuthConfig? authConfig;
  MoodishUser? user;
  List<MealMemoryEntry> mealMemory = [];
  HealthStatus? health;

  ThemeMode themeMode = ThemeMode.system;

  bool get isLoggedIn => user != null;

  Future<void> loadTheme() async {
    final prefs = await SharedPreferences.getInstance();
    final saved = prefs.getString('moodish.theme');
    themeMode = switch (saved) {
      'light' => ThemeMode.light,
      'dark' => ThemeMode.dark,
      _ => ThemeMode.system,
    };
    notifyListeners();
  }

  Future<void> setTheme(ThemeMode mode) async {
    themeMode = mode;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('moodish.theme', switch (mode) {
      ThemeMode.light => 'light',
      ThemeMode.dark => 'dark',
      ThemeMode.system => 'system',
    });
    notifyListeners();
  }

  Future<void> bootstrap() async {
    await api.restoreSession();
    const maxAttempts = 6;
    var attempt = 1;
    while (attempt <= maxAttempts) {
      phase = BootstrapPhase.connecting;
      connectingAttempt = attempt;
      notifyListeners();
      try {
        final result = await api.bootstrap();
        authConfig = result.config;
        user = result.user;
        mealMemory = result.mealMemory;
        health = result.health;
        phase = BootstrapPhase.ready;
        notifyListeners();
        return;
      } catch (_) {
        if (attempt == maxAttempts) {
          phase = BootstrapPhase.failed;
          failureMessage = 'Moodish is taking longer than usual to wake up. Pull to retry.';
          notifyListeners();
          return;
        }
        final delayMs = (300 * (1 << attempt)).clamp(0, 5000);
        await Future.delayed(Duration(milliseconds: delayMs));
        attempt += 1;
      }
    }
  }

  Future<void> loginWithDemo() async {
    final loggedInUser = await api.demoLogin();
    user = loggedInUser;
    notifyListeners();
  }

  Future<void> logout() async {
    await api.logout();
    user = null;
    mealMemory = [];
    notifyListeners();
  }
}
